packages/core/test/acceptance/di_spec.ts

```diff
@@ -739,6 +716,742 @@ describe('di', () => {
+             () => {
+               @Component({
+                 selector: 'parent',
+                 template: '<child></child>',
+                 providers: [{
+                   provide: 'token',
+                   useValue: 'PARENT',
+                 }]
+               })
+               class ParentComponent {
+               }
+
+               @Component({
+                 selector: 'child',
+                 template: '...',
+                 providers: [{
+                   provide: 'token',
+                   useValue: 'CHILD',
+                 }]
+               })
+               class ChildComponent {
+                 constructor(@Host() @SkipSelf() public injector: Injector) {}
+               }
+
+               TestBed.configureTestingModule({
+                 declarations: [ParentComponent, ChildComponent],
+               });
+
+               // Ivy has different error message when dependency is not found
+               const expectedErrorMessage = ivyEnabled ?
+                   /NG0201: No provider for Injector found in NodeInjector/ :
+                   /No provider for Injector/;
+               expect(() => TestBed.createComponent(ParentComponent))
+                   .toThrowError(expectedErrorMessage);
+             });
+
+          it('should not throw when injecting Injectors using @SkipSelf, @Host, and @Optional and no Injectors are available in a current view',
+             () => {
+               @Component({
+                 selector: 'parent',
+                 template: '<child></child>',
+                 providers: [{
+                   provide: 'token',
+                   useValue: 'PARENT',
+                 }]
+               })
+               class ParentComponent {
+               }
+
+               @Component({
+                 selector: 'child',
+                 template: '...',
+                 providers: [{
+                   provide: 'token',
+                   useValue: 'CHILD',
+                 }]
+               })
+               class ChildComponent {
+                 constructor(@Host() @SkipSelf() @Optional() public injector: Injector) {}
+               }
+
+               TestBed.configureTestingModule({
+                 declarations: [ParentComponent, ChildComponent],
+               });
+
+               // Ivy has different error message when dependency is not found
+               const expectedErrorMessage = ivyEnabled ?
+                   /NG0201: No provider for Injector found in NodeInjector/ :
+                   /No provider for Injector/;
+               expect(() => TestBed.createComponent(ParentComponent))
+                   .not.toThrowError(expectedErrorMessage);
+             });
+        });
+
+        describe('ElementRef', () => {
+          it('should lookup module injector in case @SkipSelf is used for `ElementRef` token and Component has no parent',
```

Comment:
I think we can add a quick comment here (and in similar tests for other special token) to indicate that this behavior is here only for legacy reasons. We can also put a similar comment to the `di.ts`.

Proposed comment for the test:
```suggestion
          // While tokens like `ElementRef` make sense only in a context of a NodeInjector, ViewEngine
          // also used `ModuleInjector` tree to lookup such tokens. In Ivy we replicate this behavior for now to avoid
          // breaking changes.
          it('should lookup module injector in case @SkipSelf is used for `ElementRef` token and Component has no parent',
```

Reply:
Added the comments in several places to clarify why it's there.